[XLA] Allow the tuple simplifier to operate on only subcomputations

DavidNorman opened the pull request:
This change optionally allows for the tuple simplifier to exclude the entry computation from its processing.

DavidNorman committed:
Allow the tuple simplifier to operate on only subcomputations

DavidNorman committed:
Remove unnecessary trace

drpngx commented:
@jlebar could you take a look? Feel free to suggest another reviewer otherwise.

jlebar commented on tensorflow/compiler/xla/service/tuple_simplifier.cc:
```diff
@@ -30,10 +30,19 @@ limitations under the License.
   // Initially add all GTE and Tuple instructions to the worklist.
   std::queue<HloInstruction*> worklist;
   for (auto* computation : module->computations()) {
+    LOG(INFO) << "*** Computation is " << computation->name() << " at " << (void*)computation;
```
Please use VLOG(N) instead of LOG(INFO).  Although I'm not sure these are actually useful to keep at all?  If you want to keep them, please try to follow the style in other passes wrt e.g. the `***` headers and so on.

jlebar commented on tensorflow/compiler/xla/service/tuple_simplifier.h:
```diff
@@ -27,13 +27,17 @@ namespace xla {
 class TupleSimplifier : public HloPassInterface {
  public:
-  TupleSimplifier() {}
+  TupleSimplifier() : TupleSimplifier(false) {}
```
`TupleSimplifier(/*exclude_entry_computation=*/false)`

jlebar commented on tensorflow/compiler/xla/service/tuple_simplifier.h:
```diff
@@ -27,13 +27,17 @@ namespace xla {
  public:
-  TupleSimplifier() {}
+  TupleSimplifier() : TupleSimplifier(false) {}
+  TupleSimplifier(bool exclude_entry_computation);
```
`explicit`

jlebar replied:
Please add a brief comment explaining what is the meaning of `exclude_entry_computation` (preferably without just repeating the words in the variable name :).

jlebar reviewed (changes requested):
I presume you want this for your out-of-tree backend?

That's totally fine by me, but can you please add tests and a comment indicating that it's used out-of-tree?  Otherwise I'm certain an enterprising Googler is going to look at this, realize that this is dead code, and rip it out.

DavidNorman commented:
thanks - will do all those.  that LOG(INFO) was debug left in by mistake.

DavidNorman committed:
Add a test for the tuple simplifier

Summary: Adding a test for the tuple simplifier following review of public Pull Request

Test Plan: ran this specific test, and all existing poplar tests

Reviewers: jamesn

Reviewed By: jamesn

Differential Revision: https://phabricator.sourcevertex.net/D4548

DavidNorman committed:
Merge remote-tracking branch 'tensorflow/master' into optionally-dont-process-entry-computation

DavidNorman committed:
Add comment to the parameter in the default constructor

DavidNorman commented:
Updated

jlebar commented on tensorflow/compiler/xla/service/tuple_simplifier_test.cc:
```diff
@@ -211,5 +217,76 @@ TEST_F(TupleSimplifierTest, IncompatibleTuples) {
+    entry = module->AddEntryComputation(builder.Build());
+  }
+
+  Run(module.get(), /*change_expected=*/true, /*exclude_entry*/ true);
```
`/*exclude_entry=*/`

(We do this because clang-tidy has automated checks that ensure that the comment matches the param name.)

DavidNorman replied:
i will fix.  well spotted

jlebar commented on tensorflow/compiler/xla/service/tuple_simplifier_test.cc:
```diff
@@ -211,5 +217,76 @@ TEST_F(TupleSimplifierTest, IncompatibleTuples) {
   EXPECT_THAT(computation->root_instruction(), tuple);
 }
 
+TEST_F(TupleSimplifierTest, CanExcludeEntryComputation) {
```
No need to change it, how you have it here is totally fine, but FYI most of us enjoy writing tests using the textual format.  (Update your tree and grep for `ParseHloString`.)

DavidNorman replied:
i have some tests which use that text string processing thing, but I'm not sure how you can refer to the pre-optimized instructions at post-optimization check time.   

i suppose you could walk up the tree from the root and check the opcode or metadata.

jlebar replied:
> i suppose you could walk up the tree from the root and check the opcode or metadata.

Yes, see hlo_matchers.h.

jlebar commented on tensorflow/compiler/xla/service/tuple_simplifier_test.cc:
```diff
@@ -211,5 +217,76 @@ TEST_F(TupleSimplifierTest, IncompatibleTuples) {
 }
 
+TEST_F(TupleSimplifierTest, CanExcludeEntryComputation) {
+  //  Verify that the root computation can be excluded
```
Nit, move this comment up one line?

DavidNorman replied:
the other tests in the file seem to have their comments following the test declaration.  i think it is best to leave it here.

DavidNorman committed:
Correct clang-tidy linting issue

DavidNorman commented:
update done.  thanks :)

jlebar commented on tensorflow/compiler/xla/service/tuple_simplifier_test.cc:
```diff
@@ -211,5 +217,76 @@ TEST_F(TupleSimplifierTest, IncompatibleTuples) {
+    entry = module->AddEntryComputation(builder.Build());
+  }
+
+  Run(module.get(), /*change_expected=*/true, /*exclude_entry=*/ true);
```
Our style is no space between end of the comment and `true`.

(Sorry...  :-/  I believe clang-format will catch this, and it looks like we have open-sourced a `.clang-format` file for XLA.  If you use clang-format it's very important to configure it only to format the lines you actually changed.  Otherwise you'll have tons of spurious whitespace changes to the file -- its format is not stable.  The git-clang-format script in LLVM is one way to do this.  https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format)

DavidNorman commented:
Does someone have to tell the CI system run run the unit tests, so that it will be merged?

jlebar commented:
@drpngx can you help us get this committed?

drpngx commented:
Sure, kicking off the tests...

drpngx commented:
Merged. Thank you for the contribution!
